README.md
# timmy - Use some TLS servers for trusted timestamping
0.1.0-RC1

"Trusted timestamping" is using a trusted party to sign a timestamp together with a piece of data (normally a hash of a document). This can for example be used to prove that a document was authored at, or before, a certain date.

`timmy` is a command line tool that uses a peculiarity in the TLS protocol to let many TLS servers act as a trusted timestamping server. That is, you can let a TLS server, such as www.google.com, sign your hash with a *timestamp provided by the server*.

This program is a **proof of concept**. Use this program at your own risk. See the About section for more details.

## Usage

To sign a document you do:

    $ timmy --sign document > output.json
	www.google.com signed SHA-256 bf921b493168a... at
	  2015-08-22T20:04:24Z (Unix Timestamp: 1440273864).
	  Certificate used has SHA-1 fingerprint 07f8e9d3f...
	  and expires 2015-11-10 00:00:00 UTC.

To verify a document you do:

	$ sha256sum document
	bf921b493168a...  document
    
    $ timmy --verify output.json
    Signature verification SUCCESS.
    Warning! Signature only verified against first X509 certificate.
    Please verify yourself that the certificate chain is valid.
    
    Certificate Subject: C=US/ST=California/L=Mountain View/O=Google Inc/CN=google.com/
    Certificate SHA-1 Fingerprint: 07f8e9d3fdf0135583a07015a208731d213f7de0
    Signed Time: 2015-08-22 20:04:24 UTC (timestamp 1440273864)
    Signed User-supplied SHA-256 Hash: bf921b493168a050884f723dc13fa4b1fc0afe95d06ce8cd4d66b4087204cbd0

### Gotchas

By default the program will use www.google.com:443 for signing, which as of writing (August 2015) work correctly for our purpose. Some TLS servers may use implementations that do not include a valid timestamp, [such as recent versions of OpenSSL](http://git.openssl.org/gitweb/?p=openssl.git;a=commitdiff;h=2016265dfbab162ec30718b5e7480add42598158). RFC compliant implementations should include a timestamp but it's not required to be valid. See chapter 7.4.1.2 of RFC 5426 or RFC 2246 for details. Here's a short list of servers that work:

* www.google.com:443
* www.nsa.gov:443
* letsencrypt.org:443
* www.symantec.com:443
* www.godaddy.com:443

timmy will detect "invalid" servers and refuse to sign (TODO: implement similar for verification):

    $ timmy --sign document --host facebook.com
	ERROR! Server responded with invalid time! Aborting.

A further gotcha is that many of the *valid* servers are configured with short lived certificates nowadays. For example Googles certificates are routinely only valid for a few months at most. On signature verification timmy will check that the timestamp signed falls within the valid dates. You may see this after a while:

    $ timmy --verify output.json
    ERROR! Signature verification FAILURE: Certificate has expired.

Yet another gotcha is of course that a certificate may have been revoked. timmy will not check this for you.

### Attacks

As mentioned above, not all TLS servers supply valid timestamps. This is true for many big websites, such as facebook.com or amazon.com. An attacker can abuse this by repeatedly sending requests to any of these servers until a desired (within some range) timestamp is returned. You should only accept signatures from servers you know are trustworthy in this regard.

As a further complication, it could happen that a server used to reply with random timestamps, but sometimes in the future changes this behavior, or vice versa. This may cast doubt on the signatures, especially if you are not sure about a servers behavior at the specific point in time when the data was signed.

Use timmy at your own risk.

### Input, Output, Signature and Certificates

When you sign a document with timmy, you sign a SHA-256 hash. For convenience timmy will calculate this hash for you for your input file. Lets call this hash `H`. `H` is sent to the server in a special way and you will get some data back.

The output when you sign is a JSON blob that contains three Base64 coded fields:

* `"certificates"` is a list of certificates: the certificate chain returned by the TLS server. The first certificate in the list is the one that signs your data.
* `"blob"` is the data signed. The first 32 bytes is the hash `H` and the following 4 bytes is the timestamp. The rest of the data in the blob are internal to the TLS handshake and can be ignored for our purpose.
* `"signature"` is the server's signature of the blob. It is signed by the private key corresponding to the public key in `certificates[0]`.

## Building

This program is written in Rust. It has been tested with the 1.2.0 release.

    $ cargo build [ --release ]

There is also a debian package you can build using normal means.

## Appendix A: Signature Details

To avoid confusion I'd only recommend you to read this section if you are comfortable with cryptography and the TLS protocol. Otherwise this may be confusing.

The "signature" returned by the TLS server is a PKCS 1.5 padded hash of the "blob". The hash in this case is the unusual TLS 1 construction TLSHash(h) = MD5(h) || SHA1(h).

## About

The author (Björn) discovered this curiosity of the TLS protocol back in 2012, and wrote a blog post about it here: [blog.bjrn.se](http://blog.bjrn.se/2012/07/fun-with-tls-handshake.html). Now three years later he wanted to learn the Rust programming language and resumed the project. The author is not endorsing using this program for real trusted timestamping needs: use at your own risk.

See LICENSE for licensing information. Copyright (C) Björn Edström <be@bjrn.se> 2012, 2015.


debian/changelog
timmy (0.1.0-RC1) unstable; urgency=low

  * Initial Release.

 -- Björn Edström <be@bjrn.se>  Sun, 23 Aug 2015 16:55:00 +0100


src/main.rs
// Copyright (c) Björn Edström <be@bjrn.se> 2015.
// See LICENSE for details.

extern crate chrono;
extern crate crypto;
extern crate getopts;
extern crate num;
extern crate rustc_serialize;

mod timmy;

use chrono::datetime::DateTime;
use chrono::offset::utc::UTC;
use crypto::digest::Digest;
use crypto::sha2::Sha256;
use getopts::Options;
use num::bigint::BigUint;
use rustc_serialize::base64::{STANDARD, ToBase64, FromBase64};
use rustc_serialize::json::{self};
use std::env;
use std::fs::File;
use std::io::Read;
use std::io::Write;
use std::io;
use std::net::TcpStream;
use std::process::exit;

use timmy::rsa::*;
use timmy::util::*;
use timmy::tls::*;
use timmy::x509::*;

macro_rules! println_stderr(
    ($($arg:tt)*) => (
        match writeln!(&mut ::std::io::stderr(), $($arg)* ) {
            Ok(_) => {},
            Err(x) => panic!("Unable to write to stderr: {}", x),
        }
    )
);

#[derive(RustcDecodable, RustcEncodable, Debug)]
struct JsonOutput {
    blob: String,
    certificates: Vec<String>,
    signature: String
}

fn verify(blob: &Vec<u8>, cert0: &Vec<u8>, signature: &Vec<u8>) -> i32 {
    // Parse certificates
    let cert = X509Certificate::new(cert0.clone());
    let parsed_cert = match cert.parse() {
        Ok(cert) => cert,
        Err(e) => {
            println_stderr!("ERROR! Failed to parse certificate: {:?}", e);
            return 1;
        }
    };

    // Check if certificate expired.
    let utc_now: DateTime<UTC> = UTC::now();
    let cert_still_valid = utc_now >= parsed_cert.validity.0 && utc_now <= parsed_cert.validity.1;
    if !cert_still_valid {
        println_stderr!("ERROR! Signature verification FAILURE: Certificate has expired.");
        return 1;
    }

    match parsed_cert.key {
        PublicKey::RSA(rsa_n, rsa_e) => {

            // Perform RSA signature operation
            let signature_int = BigUint::from_bytes_be(&signature);
            let sig_op = rsa_encrypt(&signature_int, &rsa_e,  &rsa_n);

            // Fix up PKCS1.5 blob
            let mut raw_pkcs1_ = sig_op.to_bytes_be();
            raw_pkcs1_.insert(0, 0x00);
            let raw_pkcs1 = raw_pkcs1_;

            // Construct our own PKCS 1.5 blob to compare against
            let tls_hash = weird_tls_hash(&blob);
            let constructed_pkcs1 = make_pkcs1_sig_padding(&rsa_n, &tls_hash);

            let valid_signature = crypto_compare(&constructed_pkcs1, &raw_pkcs1);

            if valid_signature {
                // Signature is valid, parse out fields from blob.
                let mut tls = TLSHandshake::new();
                tls.client_random = blob[0..32].iter().cloned().collect();
                tls.server_random = blob[32..64].iter().cloned().collect();

                // Validate signed timestamp against certificate validity period.
                let unix_timestamp = tls.get_unix_timestamp();
                let ts = timestamp_to_datetime(unix_timestamp);
                let valid_dates = ts >= parsed_cert.validity.0 && ts <= parsed_cert.validity.1;

                if valid_dates {
                    // Success!
                    println!("Signature verification SUCCESS.");
                    println!("Warning! Signature only verified against first X509 certificate.");
                    println!("Please verify yourself that the certificate chain is valid.");
                    println!("");
                    println!("Certificate Subject: {}", parsed_cert.subject);
                    println!("Certificate SHA-1 Fingerprint: {}", to_hex_string(&cert.sha1_fingerprint().unwrap()));
                    println!("Signed Time: {} (timestamp {})", ts, unix_timestamp);
                    println!("Signed User-supplied SHA-256 Hash: {}", to_hex_string(&tls.client_random));
                } else {
                    println!("ERROR! Signature verification FAILURE: Invalid timestamp.");
                    return 1;
                }

            } else {
                println!("ERROR! Signature verification FAILURE.");
                return 1;
            }
        }
    }

    0
}

fn sign(server: &String, port: &u16, hash_buf: &[u8; 32]) -> i32 {
    // Construct our special ClientHello message and send it to the server.
    let conn_str = format!("{}:{}", server, port);
    let mut tcpconn = match TcpStream::connect(&conn_str[..]) {
        Ok(conn) => conn,
        Err(err) => {
            println_stderr!("ERROR! Failed to connect to server: {}", err);
            return 1;
        }
    };

    let mut ch = SimpleBinaryWriter::new();
    create_special_client_hello(&mut ch, &hash_buf);

    match tcpconn.write(&ch.buf) {
        Err(_) => {
            println_stderr!("ERROR! Failed to write to server.");
            return 1;
        }
        Ok(_) => {}
    }

    // Parse the response.
    let mut pars = BinaryParser::new(&mut tcpconn);
    let mut tls = TLSHandshake::new();

    tls.client_random.extend(hash_buf[0..32].iter());

    match tls.parse(&mut pars) {
        Ok(_) => {},
        Err(s) => {
            println_stderr!("ERROR! {}", s);
            exit(1);
        }
    }

    // Extract timestamp and perform checks on it.
    let unix_timestamp = tls.get_unix_timestamp();
    let ts = timestamp_to_datetime(unix_timestamp);

    let utc_now: DateTime<UTC> = UTC::now();
    let utc_now_ts = utc_now.timestamp() as u32;

    if unix_timestamp < utc_now_ts - 3600 || unix_timestamp > utc_now_ts + 3600 {
        println_stderr!("ERROR! Server responded with invalid time! Aborting.");
        return 1;
    }

    // Parse certificate that signed.
    let cert = X509Certificate::new(tls.certs[0].buf.clone());
    let parsed_cert = match cert.parse() {
        Ok(cert) => cert,
        Err(e) => {
            println_stderr!("ERROR! Failed to parse certificate: {:?}", e);
            return 1;
        }
    };

    // Output banner and result JSON blob.
    println_stderr!("{} signed SHA-256 {} at {:?} (Unix Timestamp: {}). Certificate used has SHA-1 fingerprint {} and expires {}.",
                    server, to_hex_string(&hash_buf[0..32].iter().cloned().collect()),
                    ts, unix_timestamp, to_hex_string(&tls.certs[0].sha1_fingerprint().unwrap()), parsed_cert.validity.1);

    let json_output = JsonOutput {
        blob: tls.signed_blob.to_base64(STANDARD),
        signature: tls.signature.to_base64(STANDARD),
        certificates: tls.certs.iter().map(
            |c| c.buf.to_base64(STANDARD)
            ).collect(),
    };

    match json::encode(&json_output) {
        Ok(json) => {
            println!("{}", json);
        },
        Err(_) => {
            println_stderr!("ERROR! Failed to serialize json.");
            return 1;
        }
    }

    0
}

fn print_usage(program: &str, opts: Options) {
    let brief = format!("Usage: {} [options]", program);
    print!("{}", opts.usage(&brief));
}

fn main() {
    let args: Vec<String> = env::args().collect();
    let program = args[0].clone();
    let mut opts = Options::new();

    opts.optopt("V", "verify", "verify signature object", "PATH");
    opts.optopt("S", "sign", "input file to sign [-]", "PATH");
    opts.optopt("H", "host", "set server for signing [www.google.com]", "HOSTNAME");
    opts.optopt("p", "port", "set port for --server [443]", "PORT");

    opts.optflag("h", "help", "print this help menu");

    let parse_result = opts.parse(&args[1..]);

    if let Err(f) = parse_result {
        println_stderr!("ERROR: {}", f.to_string());
        return;
    }

    let matches = parse_result.unwrap();

    if matches.opt_present("h") {
        print_usage(&program, opts);
        return;
    }

    // parse "host"
    let server = if matches.opt_present("H") {
        matches.opt_str("H").unwrap()
    } else {
        "www.google.com".to_string()
    };

    // parse "port"
    let port: u16 = if matches.opt_present("p") {
        let port_str = matches.opt_str("p").unwrap();
        match port_str.parse() {
            Ok(p) => p,
            Err(_) => {
                print_usage(&program, opts);
                return;
            }
        }
    } else {
        443
    };

    if !matches.opt_present("V") {
        // sign

        // parse "file" and hash it
        let mut hasher = Sha256::new();

        if matches.opt_present("S") {
            let path = matches.opt_str("S").unwrap();
            let mut file_handle = match File::open(path) {
                Ok(f) => f,
                Err(s) => {
                    println_stderr!("Reading file failed: {}", s);
                    exit(1);
                }
            };

            hash_content(&mut file_handle, &mut hasher);
        } else {
            hash_content(&mut io::stdin(), &mut hasher);
        };

        let mut hash_buf: [u8; 32] = [0; 32];
        hasher.result(&mut hash_buf);

        let ret = sign(&server, &port, &hash_buf);

        exit(ret);
    } else {
        // verify

        let path = matches.opt_str("V").unwrap();
        let mut contents: Vec<u8> = Vec::new();

        let mut file_handle = match File::open(path) {
                Ok(f) => f,
                Err(s) => {
                    println_stderr!("Reading file failed: {}", s);
                    exit(1);
                }
            };

        match file_handle.read_to_end(&mut contents) {
            Err(s) => {
                println_stderr!("Reading file failed: {}", s);
                exit(1);
            }
            Ok(_) => {}
        }

        let filestr = String::from_utf8(contents).unwrap();

        let json_blob: JsonOutput = match json::decode(&filestr) {
            Ok(json) => { json },
            Err(_) => {
                println_stderr!("ERROR! Failed to deserialize json.");
                exit(1);
            }
        };

        let blob = match json_blob.blob.from_base64() {
            Ok(res) => { res },
            Err(_) => {
                println_stderr!("ERROR! Failed to decode base64 for: blob");
                exit(1);
            }
        };

        let cert0 = match json_blob.certificates[0].from_base64() {
            Ok(res) => { res },
            Err(_) => {
                println_stderr!("ERROR! Failed to decode base64 for: certificates[0]");
                exit(1);
            }
        };

        let signature = match json_blob.signature.from_base64() {
            Ok(res) => { res },
            Err(_) => {
                println_stderr!("ERROR! Failed to decode base64 for: signature");
                exit(1);
            }
        };

        let ret = verify(&blob, &cert0, &signature);

        exit(ret);
    }
}
